### ISSUE TITLE
A few questions about  the algorithm details.

### ISSUE THREAD

**Issue by my-yy** [swesterfeld/audiowmark#31]
Thank you for your awesome library, which provides great ease of use.  I am interested in the principle behind it, and after reading the code, I have three questions:

1. I notice the bit information is embedded by adjusting amplitude:

```
const float mag = abs(fft_data[i]);
if (mag > min_mag) {
    const float mag_factor = powf(mag, -Params::water_delta * data_bit_sign);
    fft_delta_spect[i] = fft_data[i] * (mag_factor - 1);
}
```
Is the position of ascending and descending chosen randomly?

2.  In this [issue](https://github.com/swesterfeld/audiowmark/issues/28) you said "if (umag > dmag) we decode an 1 bit, otherwise 0." 
The `mag_factor` is generated without considering the difference between umag and dmag.  Can the adjustment be sufficient to ensure the encoding of the desired bit?
For example, if we want to code bit 1, but currently umg < dmag,  after the adjustment, can we make sure that umg > dmag ?

6. Block A and B contain complete information, and blocks A-B can be combined for further error correction. This is interesting. What is the principle of this combined error correction? Could you please provide some reference materials?

Thank you for your reply.

**Comment by swesterfeld** [comment]
**Question 1**
Yes, the positions are chosen (pseudo) randomly depending on the key.

**Question 2**
Well, it can happen that after the adjustment the desired effect is not accomplished, but the error correction will take care of this. Note that even if we have successfully encoded the bit, altering the signal by something like ogg encoding the audio can still flip single bits. As long as it is not too bad, error correction will handle this, too.

There are two general ideas I have for improving audiowmark here. 
- we could try to ensure that `mag_factor` is increased enough to always ensure that the bit is encoded strong enough so that decoding will always succeed (at least on an unaltered signal)
- we could also use a psychoacoustic model (like mp3/ogg) to select which bands are less likely to be annoying for the listener when altered, and encode more information in these bands

When implementing one (or a combination of both) one would have to do testing to ensure that this really improves robustness at the same level of subjective quality.

**Question 3**
I don't have any reference material for this strategy. Basically I just use a convolutional code with double the order. Then if I only have one block (A or B) I set the influence of the bits I don't know to zero for soft decoding. So the code will only use the bits of the block I have. I tested that this really improves error correction if both blocks are present. I don't know if this approach is theoretically optimal.
